File: scripts/strip-url-php.js
const fs = require('fs');
const path = require('path');

const root = path.join(__dirname, '..');
const dirs = [root, path.join(root, 'admin'), path.join(root, 'client'), path.join(root, 'includes')];
const seen = new Set();
let changed = 0;

function strip(content) {
  let n = content;
  // .php before PHP echo in attributes: href="/path.php<?=
  n = n.replace(/(\b(?:href|action|formaction)\s*=\s*["'])(\/(?!api\/)[^"'<]*?)\.php(<\?=)/gi,
    '$1$2$3');
  // .php before PHP echo in quoted strings in arrays: '/client/foo.php',
  n = n.replace(/(["'])(\/(?!api\/)(?:admin|client)\/[a-zA-Z0-9_/-]*?)\.php(["',])/g,
    '$1$2$3');
  // hash anchors: /page.php#section
  n = n.replace(/(\/(?!api\/)[a-zA-Z0-9_/-]+)\.php(#)/g, '$1$2');
  // form action index.php#contact
  n = n.replace(/(\/(?!api\/)[a-zA-Z0-9_/-]*)\.php(#)/g, '$1$2');
  return n;
}

for (const d of dirs) {
  if (!fs.existsSync(d)) continue;
  const stack = [d];
  while (stack.length) {
    const dir = stack.pop();
    for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, ent.name);
      if (ent.isDirectory()) {
        if (full.includes(`${path.sep}api${path.sep}`)) continue;
        stack.push(full);
        continue;
      }
      if (!ent.name.endsWith('.php')) continue;
      const norm = path.normalize(full);
      if (seen.has(norm)) continue;
      seen.add(norm);
      const content = fs.readFileSync(norm, 'utf8');
      const updated = strip(content);
      if (updated !== content) {
        fs.writeFileSync(norm, updated, 'utf8');
        changed++;
        console.log(path.relative(root, norm));
      }
    }
  }
}

console.log(`Done. ${changed} files updated.`);

File: scripts/strip-url-php.py
import re
import os

root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
dirs = [root, os.path.join(root, "admin"), os.path.join(root, "client"), os.path.join(root, "includes")]
pat = re.compile(r'\b(href|action)\s*=\s*(["\'])(/(?!api/)[^"\']*?)\.php(\?[^"\']*)?\2', re.I)
changed = 0
seen = set()

for d in dirs:
    if not os.path.isdir(d):
        continue
    for dirpath, _, files in os.walk(d):
        if os.sep + "api" + os.sep in dirpath:
            continue
        for fn in files:
            if not fn.endswith(".php"):
                continue
            path = os.path.normpath(os.path.join(dirpath, fn))
            if path in seen:
                continue
            seen.add(path)
            with open(path, "r", encoding="utf-8", errors="replace") as f:
                content = f.read()
            new = pat.sub(lambda m: f"{m.group(1)}={m.group(2)}{m.group(3)}{m.group(4) or ''}{m.group(2)}", content)
            if new != content:
                with open(path, "w", encoding="utf-8", newline="\n") as f:
                    f.write(new)
                changed += 1
                print(os.path.relpath(path, root))

print(f"Done. {changed} files updated.")
